File: java-workshop/part-4-behavior/src/test/java/service/OrderCancellationServiceTest.java
package service;

import domain.CancellationResult;
import domain.Order;
import domain.OrderItem;
import domain.OrderStatus;
import gateway.PaymentGateway;
import gateway.ShippingGateway;
import helper.InMemoryOrderRepository;
import helper.InMemoryStockRepository;
import helper.OrderBuilder;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class OrderCancellationServiceTest {

    @Mock
    private PaymentGateway paymentGateway;

    @Mock
    private ShippingGateway shippingGateway;

    private InMemoryOrderRepository orderRepository;
    private InMemoryStockRepository stockRepository;
    private OrderCancellationService service;

    @BeforeEach
    void setUp() {
        orderRepository = new InMemoryOrderRepository();
        stockRepository = new InMemoryStockRepository();

        StockService stockService = new RepositoryStockService(stockRepository);
        PaymentService paymentService = new GatewayPaymentService(paymentGateway);
        ShippingService shippingService = new GatewayShippingService(shippingGateway);

        service = new OrderCancellationService(orderRepository, paymentService, stockService, shippingService);
    }

    @Test
    void cancelOrder_succeeds_whenPaymentRefundedAndShipmentCancelled() {
        Order order = OrderBuilder.confirmed()
            .withId("ORDER-001")
            .withTransaction("TXN-123")
            .withTracking("TRACK-456")
            .withItems(List.of(new OrderItem("PROD-001", 2, 25.0)))
            .build();

        givenOrderExists(order);
        givenStockAvailable("PROD-001", 10);
        givenRefundSucceeds();
        givenShipmentCancelled();

        CancellationResult result = service.cancelOrder("ORDER-001");

        assertSuccess(result);
        verify(paymentGateway).refund("TXN-123");
        verify(shippingGateway).cancelShipment("TRACK-456");
        assertEquals(OrderStatus.CANCELLED, orderRepository.findById("ORDER-001").status());
    }

    @Test
    void cancelOrder_succeeds_whenNoShipmentExists() {
        Order order = OrderBuilder.confirmed()
            .withId("ORDER-002")
            .withTransaction("TXN-456")
            .withoutTracking()
            .withItems(List.of(new OrderItem("PROD-001", 1, 30.0)))
            .build();

        givenOrderExists(order);
        givenStockAvailable("PROD-001", 5);
        givenRefundSucceeds();

        CancellationResult result = service.cancelOrder("ORDER-002");

        assertSuccess(result);
        verify(paymentGateway).refund("TXN-456");
        assertEquals(OrderStatus.CANCELLED, orderRepository.findById("ORDER-002").status());
    }

    @Test
    void cancelOrder_fails_whenRefundFails() {
        Order order = OrderBuilder.confirmed()
            .withId("ORDER-001")
            .withTransaction("TXN-123")
            .withoutTracking()
            .withItems(List.of(new OrderItem("PROD-001", 1, 30.0)))
            .build();

        givenOrderExists(order);
        givenRefundFails();

        CancellationResult result = service.cancelOrder("ORDER-001");

        assertFailure(result, "Refund failed");
        assertEquals(OrderStatus.CONFIRMED, orderRepository.findById("ORDER-001").status());
    }

    @Test
    void cancelOrder_fails_whenOrderDoesNotExist() {
        CancellationResult result = service.cancelOrder("ORDER-UNKNOWN");

        assertFailure(result, "Order not found");
    }

    private void givenOrderExists(Order order) {
        orderRepository.with(order);
    }

    private void givenStockAvailable(String productId, int quantity) {
        stockRepository.withAvailableStock(productId, quantity);
    }

    private void givenRefundSucceeds() {
        when(paymentGateway.refund(anyString())).thenReturn(true);
    }

    private void givenRefundFails() {
        when(paymentGateway.refund(anyString())).thenReturn(false);
    }

    private void givenShipmentCancelled() {
        when(shippingGateway.cancelShipment(anyString())).thenReturn(true);
    }

    private void assertSuccess(CancellationResult result) {
        assertEquals(new CancellationResult.Success(), result);
    }

    private void assertFailure(CancellationResult result, String reason) {
        assertEquals(new CancellationResult.Failure(reason), result);
    }
}

File: csharp-workshop/Part4.Behavior.Tests/Service/OrderCancellationServiceTest.cs
using NSubstitute;
using Part4.Behavior.Domain;
using Part4.Behavior.Gateway;
using Part4.Behavior.Service;
using Part4.Behavior.Tests.Helper;
using Xunit;

namespace Part4.Behavior.Tests.Service;

public class OrderCancellationServiceTest
{
    private readonly IPaymentGateway _paymentGateway = Substitute.For<IPaymentGateway>();
    private readonly IShippingGateway _shippingGateway = Substitute.For<IShippingGateway>();
    private readonly InMemoryOrderRepository _orderRepository;
    private readonly InMemoryStockRepository _stockRepository;
    private readonly OrderCancellationService _service;

    public OrderCancellationServiceTest()
    {
        _orderRepository = new InMemoryOrderRepository();
        _stockRepository = new InMemoryStockRepository();

        IStockService stockService = new RepositoryStockService(_stockRepository);
        IPaymentService paymentService = new GatewayPaymentService(_paymentGateway);
        IShippingService shippingService = new GatewayShippingService(_shippingGateway);

        _service = new OrderCancellationService(_orderRepository, paymentService, stockService, shippingService);
    }

    [Fact]
    public void P4_CancelOrder_Succeeds_WhenPaymentRefundedAndShipmentCancelled()
    {
        var order = OrderBuilder.Confirmed()
            .WithId("ORDER-001")
            .WithTransaction("TXN-123")
            .WithTracking("TRACK-456")
            .WithItems(new List<OrderItem> { new("PROD-001", 2, 25.0) })
            .Build();

        GivenOrderExists(order);
        GivenStockAvailable("PROD-001", 10);
        GivenRefundSucceeds();
        GivenShipmentCancelled();

        CancellationResult result = _service.CancelOrder("ORDER-001");

        AssertSuccess(result);
        _paymentGateway.Received(1).Refund("TXN-123");
        _shippingGateway.Received(1).CancelShipment("TRACK-456");
        Assert.Equal(OrderStatus.Cancelled, _orderRepository.FindById("ORDER-001")!.Status);
    }

    [Fact]
    public void P4_CancelOrder_Succeeds_WhenNoShipmentExists()
    {
        var order = OrderBuilder.Confirmed()
            .WithId("ORDER-002")
            .WithTransaction("TXN-456")
            .WithoutTracking()
            .WithItems(new List<OrderItem> { new("PROD-001", 1, 30.0) })
            .Build();

        GivenOrderExists(order);
        GivenStockAvailable("PROD-001", 5);
        GivenRefundSucceeds();

        CancellationResult result = _service.CancelOrder("ORDER-002");

        AssertSuccess(result);
        _paymentGateway.Received(1).Refund("TXN-456");
        Assert.Equal(OrderStatus.Cancelled, _orderRepository.FindById("ORDER-002")!.Status);
    }

    [Fact]
    public void P4_CancelOrder_Fails_WhenRefundFails()
    {
        var order = OrderBuilder.Confirmed()
            .WithId("ORDER-001")
            .WithTransaction("TXN-123")
            .WithoutTracking()
            .WithItems(new List<OrderItem> { new("PROD-001", 1, 30.0) })
            .Build();

        GivenOrderExists(order);
        GivenRefundFails();

        CancellationResult result = _service.CancelOrder("ORDER-001");

        AssertFailure(result, "Refund failed");
        Assert.Equal(OrderStatus.Confirmed, _orderRepository.FindById("ORDER-001")!.Status);
    }

    [Fact]
    public void P4_CancelOrder_Fails_WhenOrderDoesNotExist()
    {
        CancellationResult result = _service.CancelOrder("ORDER-UNKNOWN");

        AssertFailure(result, "Order not found");
    }

    private void GivenOrderExists(Order order)
    {
        _orderRepository.With(order);
    }

    private void GivenStockAvailable(string productId, int quantity)
    {
        _stockRepository.WithAvailableStock(productId, quantity);
    }

    private void GivenRefundSucceeds()
    {
        _paymentGateway.Refund(Arg.Any<string>()).Returns(true);
    }

    private void GivenRefundFails()
    {
        _paymentGateway.Refund(Arg.Any<string>()).Returns(false);
    }

    private void GivenShipmentCancelled()
    {
        _shippingGateway.CancelShipment(Arg.Any<string>()).Returns(true);
    }

    private static void AssertSuccess(CancellationResult result)
    {
        Assert.Equal(new CancellationResult.Success(), result);
    }

    private static void AssertFailure(CancellationResult result, string reason)
    {
        Assert.Equal(new CancellationResult.Failure(reason), result);
    }
}
